### PR TITLE
ipatests: add migrate-ds integration tests for bugs.

### PR DESCRIPTION
**Pull request by menonsudhir**
ipatests: add migrate-ds integration tests for bugs.
    
    - BZ 804609: ipa user-show --all on a migrated user with a manager
      attribute no longer causes an internal error
    - BZ 753966: migrated groups with spaces in their names can be
      deleted via ipa group-del
    - BZ 809560: private groups are not created for migrated posix users
      whose GID points to an existing group
    - BZ 813389: when two groups share the same GID, migrate-ds succeeds
      with a warning instead of failing with a cryptic error
    
Added corresponding LDIF test entries (mgr_user, user_with_mgr, posixuser_no_privgrp, dupgid_group) to instance1.ldif
Added a cleanup_migrated_data() helper that removes the users and groups migrated.
    
Ref: https://pagure.io/freeipa/issue/9965
    
Assisted-by: claude <no-reply@anthropic.com>
Signed-off-by: Sudhir Menon <sumenon@redhat.com>

## Summary by Sourcery

Add integration coverage for FreeIPA directory server migration edge cases and centralize cleanup of migrated test data.

CI:
- Update temporary CI definition to run the ds_migration integration test suite with an increased timeout.

Tests:
- Add migrate-ds integration tests covering manager attribute handling, group names with spaces, private group creation for migrated POSIX users, and duplicate GID behavior.
- Introduce a reusable helper to clean up users and groups created by migrate-ds across tests.

### CODE REVIEW

**Comment by menonsudhir** [comment]
This PR can be reviewed once https://github.com/freeipa/freeipa/pull/8340 is merged.

File: ipatests/test_integration/test_ds_migration.py [hunk 1]
```diff
@@ -108,6 +117,23 @@ def install(cls, mh):
         )
         tasks.service_control_dirsrv(cls.master, "restart")
 
+    def cleanup_migrated_data(self, extra_users=None, extra_groups=None):
```
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py]: If you define the method as `def cleanup_migrated_data(self, extra_users=[], extra_groups=[]):`, you can replace the line
```
        if extra_users:
            users.extend(extra_users)
```
with a simple
```
        users.extend(extra_users)
```
and do the same for extra_groups.

File: ipatests/test_integration/test_ds_migration.py [hunk 2]
```diff
@@ -407,19 +410,72 @@ def test_bz783270_02_migrate_ds_with_compat_enabled(self):
...
+    def test_bz804609_user_show_all_no_internal_error(self):
+        """
+        bz804609: ipa user-show --all on a migrated user with a manager
+        attribute must not return an internal error.
+        https://bugzilla.redhat.com/show_bug.cgi?id=804609
+        """
+        tasks.kinit_admin(self.master)
+        pwd = self.master.config.admin_password
+
+        # --group-container=ou=nonexistent so that no groups are migrated.
```
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py]: Is it important for the test?

File: ipatests/test_integration/test_ds_migration.py [hunk 3]
```diff
@@ -407,19 +410,72 @@ def test_bz783270_02_migrate_ds_with_compat_enabled(self):
...
+                    self.ldap_uri,
+                ],
+                stdin_text=pwd,
+            )
+            result = self.master.run_command(
+                ["ipa", "user-show", "--all", "user_with_mgr"],
+            )
+            out = result.stdout_text + result.stderr_text
+            assert "an internal error has occurred" not in out.lower(), out
+            assert "mgr_user" in result.stdout_text
```
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py:434]: I would also add a check with `ipa user-show --all --raw` or a direct ldapsearch to ensure that the attribute value for manager has been properly transformed from `manager: uid=mgr_user,ou=People,dc=testrealm,dc=test` to `manager: uid=mgr_user,cn=users,cn=accounts,$BASEDN`

File: ipatests/test_integration/test_ds_migration.py [hunk 4]
```diff
@@ -506,8 +561,75 @@ def test_hashedpwd_migration(self):
...
+            )
+            result = self.master.run_command(
+                ["ipa", "group-show", "posixuser_no_privgrp"],
+                raiseonerr=False,
+            )
+            assert result.returncode != 0, (
+                "Private group should not exist for migrated posix user"
+            )
+        finally:
+            self.cleanup_migrated_data()
```
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py:606]: The test needs to cleanup user1, user2 also but it is not easy to remember to do it in each test after test_hashedpwd_migration. I would suggest to slightly modify test_hashedpwd_migration to create the data in a different users container so that it doesn't affect the other tests:
1. replace ou=People,dc=testrealm,dc=test in instance_hashedpwd.ldif with ou=HashedPwdPeople,dc=testrealm,dc=test
2. create the container ou=HashedPwdPeople,dc=testrealm,dc=test in instance_hashedpwd.ldif 
3. replace the container in the method test_hashedpwd_migration with
```
        user_container = "ou=HashedPwdPeople,{}".format(DS_BASEDN)
        group_container = "ou=HashedPwdPeople,{}".format(DS_BASEDN)
```

File: ipatests/test_integration/test_ds_migration.py [hunk 5]
```diff
@@ -506,8 +561,75 @@ def test_hashedpwd_migration(self):
...
+            out = result.stdout_text + result.stderr_text
+            assert "Expected 1" not in out, (
+                "Migration should not fail with cryptic error"
+            )
+            self.master.run_command(
+                ["ipa", "user-show", "ldapuser_0001"]
+            )
+            log = self.master.get_file_contents(
+                "/var/log/httpd/error_log", encoding="utf-8"
+            )
```
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py]: Instead of checking the log since the file creation, the test should note the log size before calling ipa migrate-ds and retrieve the new content only. Something similar to:
```
logfile= "/var/log/httpd/error_log"
logsize = tasks.get_logsize(self.master, logfile)
... call ipa migrate-ds
log = self.master.get_file_contents(logfile)[logsize:]
```

**Review by flo-renaud — COMMENTED** [review]
Hi @menonsudhir 

Please find inline comments.
Can you also update the commit message and use "Related: https://pagure.io/freeipa/issue/9965" instead of "Ref: https://pagure.io/freeipa/issue/9965"?

File: ipatests/test_integration/test_ds_migration.py (hunk 3 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py:434]: Added the change to do direct ldapsearch.

File: ipatests/test_integration/test_ds_migration.py (hunk 2 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py]: Actually not needed, removed from the test.

File: ipatests/test_integration/test_ds_migration.py (hunk 1 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py]: @frenaud the changes have been made by removing the if statement

File: ipatests/test_integration/test_ds_migration.py (hunk 4 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py:606]: 1. Replaced ou=People,dc=testrealm,dc=test in instance_hashedpwd.ldif with ou=HashedPwdPeople,dc=testrealm,dc=test
2. created the container ou=HashedPwdPeople,dc=testrealm,dc=test in instance_hashedpwd.ldif
3. replace the container in the method test_hashedpwd_migration with        
 user_container = "ou=HashedPwdPeople,{}".format(DS_BASEDN)
 group_container = "ou=HashedPwdPeople,{}".format(DS_BASEDN)

File: ipatests/test_integration/test_ds_migration.py (hunk 5 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py]: Added the change as mentioned above.

File: ipatests/test_integration/test_ds_migration.py (hunk 1 above)
Comment (flo-renaud) [reply ipatests/test_integration/test_ds_migration.py]: Did you forget to save your changes? I still see `def cleanup_migrated_data(self, extra_users=None, extra_groups=None):`

File: ipatests/test_integration/test_ds_migration.py [hunk 6]
```diff
@@ -407,19 +410,86 @@ def test_bz783270_02_migrate_ds_with_compat_enabled(self):
...
+                ["ipa", "user-show", "--all", "user_with_mgr"],
+            )
+            out = result.stdout_text + result.stderr_text
+            assert "an internal error has occurred" not in out.lower(), out
+            assert "mgr_user" in result.stdout_text
+
+            base = self.master.domain.basedn
+            user_dn = f"uid=user_with_mgr,cn=users,cn=accounts,{base}"
+            ipa_mgr_dn = f"uid=mgr_user,cn=users,cn=accounts,{base}"
+            remote_mgr_dn = f"uid=mgr_user,ou=People,{DS_BASEDN}"
```
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py]: I would make the check more strict with `remote_mgr_dn = f"manager: uid=mgr_user,ou=People,{DS_BASEDN}"`

File: ipatests/test_integration/test_ds_migration.py (hunk 1 above)
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py]: With the latest version of this PR, the method is never called with extra_users or extra_groups and could be simplified into `def cleanup_migrated_data(self)`

File: ipatests/test_integration/test_ds_migration.py [hunk 7]
```diff
@@ -506,8 +576,74 @@ def test_hashedpwd_migration(self):
...
+                "Migration should not fail with cryptic error"
+            )
+            self.master.run_command(
+                ["ipa", "user-show", "ldapuser_0001"]
+            )
+            log_bytes = self.master.get_file_contents(logfile)[logsize:]
+            log = log_bytes.decode("utf-8", errors="replace")
+            assert (
+                "should match 1 group, but it matched 2 groups" in log
+            )
```
Comment (flo-renaud) [inline ipatests/test_integration/test_ds_migration.py]: Those lines can be simplified and replaced with
```
            log_bytes = self.master.get_file_contents(logfile)[logsize:]
            assert b"should match 1 group, but it matched 2 groups" in log
```

**Review by flo-renaud — COMMENTED** [review]
Hi @menonsudhir 

please find inline comments

File: ipatests/test_integration/test_ds_migration.py (hunk 1 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py]: @flo-renaud yes actually didn't add those changes. Now its added.

File: ipatests/test_integration/test_ds_migration.py (hunk 7 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py]: This has been addressed as well.

File: ipatests/test_integration/test_ds_migration.py (hunk 1 above)
Comment (menonsudhir) [reply ipatests/test_integration/test_ds_migration.py]: Also modified the method now i.e def cleanup_migrated_data(self):

**Comment by flo-renaud** [comment]
Link to green run: [fedora-latest/temp_commit](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/a4aac65a-43a4-11f1-9d21-fa163eb9ba11)

You can remove the temp commit. Please update the commit message:
- use "Related: https://pagure.io/freeipa/issue/9965" instead of "Ref: https://pagure.io/freeipa/issue/9965"
- fix the typo in <no-reply@anth**t**opic.com>

**Comment by menonsudhir** [comment]
> fix the typo in <no-reply@anth**t**opic.com>

@flo-renaud thank you the review and adding backport labels. Removed the temp commit.

**Comment by flo-renaud** [comment]
@menonsudhir thanks for the updates. Can be merged when CI completes

**Comment by flo-renaud** [comment]
master:

* 94650026955753f2bc7aefac11fe103aa38da37c ipatests: add migrate-ds integration tests for bugs.
